src/modules/cxxpiper.cpp

```diff
@@ -0,0 +1,1223 @@
+
+#include <sys/time.h>
+#include <array>
+#include <chrono>
+#include <condition_variable>
+#include <fstream>
+#include <functional>
+#include <limits>
+#include <filesystem>
+#include <sstream>
+#include <mutex>
+#include <stdexcept>
+#include <vector>
+
+#include <espeak-ng/speak_lib.h>
+#include <onnxruntime_cxx_api.h>
+#include "json.hpp"
+#include "piper.hpp"
+#include "utf8.h"
+#include "wavfile.hpp"
+#include "../rubberband/RubberBandStretcher.h"
+
+#include "spd_audio.h"
+#include <speechd_types.h>
+#include "module_utils.h"
+
+#define MODULE_NAME     "Cxxpiper"
+#define DBG_MODNAME     "Cxxpiper"
+#define MODULE_VERSION  "0.0"
+#define DEBUG_MODULE 5
+    DECLARE_DEBUG();
+//}
+namespace cxxpiper {
+    using namespace std;
+    using json = nlohmann::json;
+    // rubberband stretcher.  "Realtime" mode, streaming (single pass, no "study"), clipping be damned.
+    // Used for real time pitch, rate and volume adjustment.
+    using RubberBand::RubberBandStretcher;
+    using namespace piper;
+
+#ifdef _CXXPIPER_VERSION
+#define _STR(x) #x
+#define STR(x) _STR(x)
+    const std::string VERSION = STR(_CXXPIPER_VERSION);
+#else
+    const std::string VERSION = "";
+#endif
+    std::string getVersion() { return VERSION; }
+const std::string instanceName{"cxxpiper"};
+
+    const float MIN_WAV_VALUE = -32766.0f;
+    const float MAX_WAV_VALUE = 32767.0f;
+
+    static const int bs = 1024;
+    static float **cbuf;
+    static piper::PiperConfig piperConfig;
+    static char *cmdInp;
+    static int stop_requested;
+
+    static size_t cbCnt = 0; // DEBUG only
+    static size_t cbTot = 0;
+    static size_t prCnt = 0;
+    static size_t prTot = 0;
+    static size_t izCnt = 0;
+    static size_t izTot = 0;
+    static size_t ezCnt = 0;
+    static size_t ezTot = 0;
+
+    extern "C" {
+	static int cxxpiper_free_cbuf();
+	static void cxxpiper_free_voice_list();
+    }
+
+// Copied from piper distribution.
+// True if the string is a single UTF-8 codepoint
+    bool isSingleCodepoint(std::string s)
+    {
+	return utf8::distance(s.begin(), s.end()) == 1;
+    }
+
+// Copied from piper distribution.
+// Get the first UTF-8 codepoint of a string
+    Phoneme getCodepoint(std::string s)
+    {
+	utf8::iterator character_iter(s.begin(), s.begin(), s.end());
+	return *character_iter;
+    }
+
+// Copied from piper distribution.
+    void parsePhonemizeConfig(json &configRoot, PhonemizeConfig &phonemizeConfig)
```

Comment:
It's quite sad that piper doesn't provide these as library functions. It'd be good to tell upstream to do so, rather than copying their code